Product Filter: Add taxonomy filter support for product queries

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   [x] I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   [x] I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   [x] Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds support for filtering products by public taxonomies using `filter_*` parameters in the main query and product collection queries. This enables taxonomy filtering alongside existing attribute filters.

- Added automatic registration of `filter_*` query vars for all public product taxonomies.
- Added `add_taxonomy_clauses()` method in `QueryClauses` class that uses efficient EXISTS subqueries.
- Uses EXISTS subqueries instead of JOINs to avoid duplicate rows and improve query performance
- Refactored `QueryClauses` to implement new `MainQueryClausesGenerator` interface for using in `MainQueryController`.
- I take this chance to refactor how we manage filter URL parameters to make it more manageable by having a single source of truth for all parameters we use.

Closes WOOPLUG-4916. Closes https://github.com/woocommerce/woocommerce/issues/59463.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add Product Filters to the Product Catalog template.
2. Enable Product Counts for all filters which have that setting.
3. On the frontend, manually add taxonomy filter parameter to the URL, verify the count and product result are correct.
  - Try different category
  - Try adding taxonomy filter param to filtered URL.

From our internal conversation, we agreed to use these params for built in taxonomy:
- `categories` for product categories.
- `tags` for product tags.
- `brands` for product brands.

For custom taxonomy, the param name format is `filter_<taxonomy_name>`. We can consider adding a filter later enabling 3PDs to change the param name of custom taxonomy, but only when we see real demand.

<details><summary>Screenshots</summary>
<p>

<img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/ccbd5fd0-144d-4972-b2af-a402478c4e4b" />
<img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/4b9a6203-d5e4-43d1-b91d-b9f52768527f" />
<img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/13446fad-1aa2-45a5-bc69-31f8fb25433d" />


</p>
</details>

dinhtungdu committed:
feat: add taxonomy filter support for product queries

Enable filtering products by public taxonomies using filter_* parameters
to support category and tag filtering in main query alongside existing
attribute filters. Uses EXISTS subqueries for efficient boolean matching
without duplicate rows.

dinhtungdu committed:
test: add product tag helper to FixtureData and comprehensive taxonomy tests

- Add get_product_tag() method to FixtureData class for consistent tag creation
- Update AbstractProductFiltersTest to use FixtureData helper instead of private method
- Add comprehensive taxonomy filter tests for both categories and tags
- Tests cover single/multiple term filtering with OR logic
- All 16 tests pass including 5 new taxonomy filter test cases

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
refactor: get all choosen taxonomy in one query

dinhtungdu committed:
fix: drop filter from param, use shorter param name

Follow Store API to use category, tag, and brand for default taxonomy.
Custom taxonomy will use rewrite slug and fallback to taxonomy name.

dinhtungdu committed:
fix: dont use rewrite slug

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -393,6 +511,77 @@ private function get_chosen_attributes( $query_vars ) {
+		}
+
+		foreach ( $this->get_taxonomy_url_params_mapping() as $taxonomy => $param ) {
+			if ( isset( $query_vars[ $param ] ) ) {
```
`$query_vars` always contains the `tag` property, as it comes from WP core, so `get_chosen_taxonomies()` is always returning `tag` as being one of the chosen taxonomies, even if it's not in the URL. Can you reproduce too?

dinhtungdu replied:
Yeah I can reproduce. We can't use existing query vars. I reverted back to original format `filter_taxonomy_name`.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -393,6 +511,77 @@ private function get_chosen_attributes( $query_vars ) {
+				),
+				array_values( $this->get_taxonomy_url_params_mapping() ),
+			)
+		);
```
Not a big deal, but I think the outer `array_values()` is redundant, no? :thinking:

dinhtungdu replied:
I made a refactoring on this. I created a class to manage filter param and it will be the single source of truth for filter params. We need to register the param as well as use them in block and query modification, so I think we need a centralized place to manage them.

Aljullu reviewed (commented):
Very nice seeing the progress here! :raised_hands: 

I found a few edge cases when testing this:

1. Filtering by `tag` only works for me when using `?product_tag=`, but not with `?tag=`. Is that expected? The other taxonomies (`category` and `brand` seem to work with the taxonomy name and the mapped name).
2. I noticed that when filtering by one category, we don't include subcategories. Is that on purpose? For example, filtering by `category=clothing` only shows me one product even though I have many products in its subcategories:

![imatge](https://github.com/user-attachments/assets/2df8d273-3265-4748-ab68-fc51e22abefc)

3. I also found out that when not finding a match, the behavior is different when using the taxonomy name or the mapped name:

`?category=no-category` | `?product_cat=no-category`
--- | ---
![imatge](https://github.com/user-attachments/assets/27995fb5-4aca-4ca9-804b-00be1ac5e1c6) | ![imatge](https://github.com/user-attachments/assets/3ab89e72-b2b3-4e3a-a9b1-1e64150ddf53)

I think ideally we should render the empty states when there are no matches. What do you think?

![imatge](https://github.com/user-attachments/assets/fa650be4-2017-4521-8669-d5147587c1dd)

---

I also think the [last Code Rabbit suggestions](https://github.com/woocommerce/woocommerce/pull/59499#pullrequestreview-2999825176) would be nice to include.

<details>
<summary>Testing environment</summary>

* WordPress 6.8.1 local dev environment
* Plugins: WooCommerce Beta Tester
* Theme: Woo Starter Theme
* Store creation: June 2025
* Store size: small
</details>

dinhtungdu committed:
feat: single source of truth for filter url parameters

dinhtungdu committed:
refactor: remove unused woocommerce_blocks_product_filters_param_keys filter hook

dinhtungdu committed:
refactor: get only public taxonomies

dinhtungdu committed:
docs: mark ProductFilters public methods as internal APIs

dinhtungdu committed:
style: fix linting errors in ProductFilters classes

dinhtungdu committed:
fix: denfense!

dinhtungdu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -218,6 +275,83 @@ public function add_attribute_clauses( $args, $chosen_attributes ) {
+			$term_ids_by_taxonomy[ $term->taxonomy ][] = $term->term_id;
+		}
+
+		foreach ( $term_ids_by_taxonomy as $taxonomy => $term_ids ) {
+			if ( empty( $term_ids ) ) {
+				continue;
+			}
+
+			$term_ids_list = '(' . implode( ',', array_map( 'absint', $term_ids ) ) . ')';
```
this is invalid, the $chosen_taxonomies is already made from public product taxonomies already.

dinhtungdu committed:
fix: filtering param value early to avoid unnecessary calculation

dinhtungdu commented:
@Aljullu thank you so much for the reviews! I addressed your comments and made a refactor to make the whole flow more predictable. Please give this one another look.

dinhtungdu committed:
refactor: remove unused method leftover from rebasing

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/Params.php:
```diff
@@ -0,0 +1,122 @@
+	/**
+	 * Get the param keys.
+	 *
+	 * @internal For exclusive usage of WooCommerce core, backwards compatibility not guaranteed.
```
Instead of adding `@internal` to each method, wouldn't it be better to mark the whole class as `@internal`? Or is there a reason not to?

dinhtungdu replied:
Oh yes, we can do that! I just didn't think about it 😂

dinhtungdu replied:
The fact that being in `Internal` directory means the class is for internal usage already.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -239,13 +304,96 @@ public function add_attribute_clauses( $args, $chosen_attributes ) {
+			)
+		);
+
+		if ( is_wp_error( $all_terms ) ) {
+			return $args;
+		}
```
Should we in some way log that there has been an error here? Even if we want everything to continue working in the frontend, `is_wp_error()` means something went wrong, no?

dinhtungdu replied:
We can use our logger. Or we can throw an exception when WP_DEBUG is turned on, which one do you prefer?

Aljullu replied:
Not a strong opinion, I lean more towards using the logger because if keeps things more similar independently of whether `WP_DEBUG` is true or not. But as mentioned, I don't have a strong opinion. :smile:

gigitux replied:
I'm okay on using WC_Logger here

dinhtungdu replied:
Looking at the implementation of the `get_terms` function, it only returns WP_Error if some of the queried taxonomies are invalid. The queried taxonomies are always valid so the error should never be returned.

Aljullu reviewed (approved):
Thanks for the updates, @dinhtungdu! This is testing great on my end. I can successfully filter by all product taxonomies using their taxonomy string and categories correctly include subcategory products.

I left a couple of minor comments inline, but not blockers, so feel free to ignore them.

dinhtungdu committed:
chore: mark the whole class as internal (instead of each method) to reduce noise

dinhtungdu committed:
fix: mark injection method as internal

dinhtungdu committed:
fix: handle the case when some selected terms are empty

dinhtungdu committed:
fix: handle invalid term

dinhtungdu committed:
feat: update param maps, use plural form for param name of built in taxonomies

dinhtungdu commented:
Since I updated the PR to change the params name, I think this PR should have another review. @gigitux as Albert is AFK and he has an overloaded review queue, I requested another review from our team and you're the chosen one!

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilters.php:
```diff
@@ -226,17 +227,7 @@ private function get_filter_params( $query_id ) {
-		 * @return array Active filters params.
-		 */
-		$filter_param_keys = array_unique( apply_filters( 'woocommerce_blocks_product_filters_param_keys', array(), array_keys( $url_query_params ) ) );
+		$filter_param_keys = wc_get_container()->get( Params::class )->get_param_keys();
```
Checking the documentation, in the `src` folder, the DI should be done in a different way: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/README.md#1-other-classes-in-the-src-directory

dinhtungdu replied:
Copying my reply from your same comment on another PR:

>  Store API (and Blocks) has its own container, so it's not possible to follow that guide right now. See https://github.com/woocommerce/woocommerce/pull/43772#discussion_r1478117593

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterAttribute.php:
```diff
@@ -28,7 +28,6 @@ final class ProductFilterAttribute extends AbstractBlock {
 	protected function initialize() {
 		parent::initialize();
 
-		add_filter( 'woocommerce_blocks_product_filters_param_keys', array( $this, 'get_filter_query_param_keys' ), 10, 2 );
```
Could this be considered a breaking change? I’m okay with proceeding, but we should probably publish a developer advisory at the very least. What do you think?

dinhtungdu replied:
I don't think so. Even it's an public filter, only us are using it and only us should be using it. The new Product Filters block is introduced in 9.9. We haven't public any documentation around that filter.

gigitux reviewed (approved):
Thanks for working on this! The PR works as expected! I left a few comments, but those are code-related. I'm pre-approving this PR.

dinhtungdu committed:
feat: log the error if any

dinhtungdu committed:
fix: filter out empty term item early

dinhtungdu committed:
fix: get correct filterable taxonomy

dinhtungdu committed:
test: add tests for Params class

Add unit tests for the Params class covering:
- Parameter key generation for all types (price, rating, status,
taxonomy, attribute)
- Parameter retrieval by type with data provider testing
- Static caching behavior verification
- Taxonomy parameter mapping for built-in taxonomies
- Custom taxonomy parameter generation with filter_ prefix
- Edge cases and error handling
- Parameter key uniqueness validation
- Public taxonomy filtering with UI requirement

dinhtungdu committed:
feat: add return type declaration to add_query_clauses method

dinhtungdu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/ParamsTest.php:
```diff
@@ -0,0 +1,259 @@
+	public function test_taxonomy_parameter_mapping() {
+		$taxonomy_params = $this->sut->get_param( 'taxonomy' );
+
+		// Test built-in taxonomy mappings.
+		$this->assertContains( 'categories', $taxonomy_params );
+		$this->assertContains( 'tags', $taxonomy_params );
+		$this->assertContains( 'brands', $taxonomy_params );
+	}
```
in `setUp`, we manually init the brand classs, so brand will always exist for the test.
